// @bugsnag/core-performance/dist/index.js
export { ResourceAttributes, SpanAttributes, attributeToJson } from './attributes.js';
export { millisecondsToNanoseconds } from './clock.js';
export { schema, validateConfig } from './config.js';
export { createClient, createNoopClient } from './core.js';
export { TracePayloadEncoder, responseStateFromStatusCode } from './delivery.js';
export { SpanEvents } from './events.js';
export { InMemoryPersistence, toPersistedPayload } from './persistence.js';
export { BufferingProcessor } from './processor.js';
export { InMemoryQueue } from './retry-queue.js';
export { default as Sampler } from './sampler.js';
export { SpanInternal, coreSpanOptionSchema, spanEndedToSpan, spanToJson } from './span.js';
export { DefaultSpanContextStorage, spanContextEquals } from './span-context.js';
export { DISCARD_END_TIME, SpanFactory } from './span-factory.js';
export { timeToNumber } from './time.js';
export { isBoolean, isLogger, isNumber, isObject, isOnSpanEndCallbacks, isParentContext, isPersistedProbability, isPlugin, isPluginArray, isSpanContext, isString, isStringArray, isStringOrRegExpArray, isStringWithLength, isTime } from './validation.js';
export { default as traceIdToSamplingRate } from './trace-id-to-sampling-rate.js';
